src/api/middleware/metrics.ts

```diff
@@ -0,0 +1,103 @@
+import express, { RequestHandler } from 'express';
+import listEndpoints from 'express-list-endpoints';
+import promBundle, { NormalizePathFn } from 'express-prom-bundle';
+import url from 'url';
+import UrlPattern from 'url-pattern';
+import { Logger } from 'winston';
+
+import { SSP_APP_NAME } from '../envars';
+
+type Options = {
+  isNormalizePathEnabled?: boolean;
+  discardUnmatched?: boolean;
+  urlPatternMaker?: (path: string) => string;
+};
+
+const captureAllRoutes = ({ urlPatternMaker }: Options, app: express.Express, logger: Logger) => {
+  const allRoutes = listEndpoints(app).filter((route) => route.path !== '/*' && route.path !== '*');
+
+  return allRoutes.map((route) => {
+    const path = route.path.endsWith('/') ? route.path.replace(/\/$/, '') : route.path;
+
+    logger.log('debug', `Route found: ${route.path} - ${JSON.stringify(route)}`);
+
+    // NOTE: urlPatternMaker has to create an UrlPattern compatible object.
+    const pattern = urlPatternMaker
+      ? urlPatternMaker(route.path)
+      : new UrlPattern(route.path, {
+          segmentNameCharset: 'a-zA-Z0-9_-',
+        }).toString();
+
+    if (!pattern) {
+      logger.log('debug', 'skipping route due to empty path');
+    }
+
+    return { ...route, path: pattern || path };
+  });
+};
+
+const makeMetricsApiMiddleware = (options: Options, logger: Logger) => {
+  const { isNormalizePathEnabled, discardUnmatched } = options;
+  const allRoutes = [] as { methods: string[]; path: string; pattern?: UrlPattern }[];
+
+  const normalizePath: NormalizePathFn = (req, _opts) => {
+    if (!isNormalizePathEnabled) {
+      return req.url;
+    }
+
+    const path = (() => {
+      const parsedPath = url.parse(req.originalUrl || req.url).pathname || req.url;
+      if (parsedPath.endsWith('/')) {
+        // Remove trailing slash
+        return parsedPath.replace(/\/$/, '');
+      }
+      return parsedPath;
+    })();
+
+    const pattern = allRoutes.filter((route) => {
+      if (route.methods.indexOf(req.method) === -1) {
+        return false;
+      }
+
+      if (route.path.match(path)) {
+        return true;
+      }
+
+      return false;
+    })[0]?.pattern;
+
+    if (discardUnmatched && !pattern) {
+      return '';
+    }
+
+    return pattern?.stringify() || path || '';
+  };
+
+  const metricsMiddleware = promBundle({
+    includeMethod: true,
+    includePath: true,
+    autoregister: false,
+    customLabels: {
+      application: SSP_APP_NAME,
+    },
+    buckets: [0.03, 0.3, 1, 1.5, 3, 5, 10],
```

Comment:
what are these buckets for? and why are using those numbers?

Reply:
>what are these buckets for?

More info [here](https://www.timescale.com/blog/four-types-prometheus-metrics-to-collect/#:~:text=Buckets%20in%20Prometheus%20are%20inclusive,than%20or%20equal%20to%20N.)
It enables the histogram metric, and gives the distribution of requests that fall in that bucket. 

>why are we using those number?

To keep them consistent with what the tcportal is publishing. The values kind of make sense to me, as these are simple apis and anything more than 10seconds is probably too high, so it gives a distribution between 0-10